### DISCUSSION TITLE
overseer.nvim component to mini.statusline

### DISCUSSION THREAD

**Discussion by dhruvinsh** [nvim-mini/mini.nvim#1200] [Q&A] [answered]
Hi all,
Does anyone have tried adding [overseer.nvim](https://github.com/stevearc/overseer.nvim) job status as custom component to `mini.statusline`?

https://github.com/stevearc/overseer.nvim/blob/master/doc/third_party.md#lualine

**Comment by echasnovski** [comment, marked as answer]
Taking [its lualine component as a reference](https://github.com/stevearc/overseer.nvim/blob/master/lua/lualine/components/overseer.lua), here is how I'd approach this:

- Create a file 'lua/overseer-stl.lua' inside your config directory. This will be your statusline component. Populate it with the following content:
    ```lua
    local constants = require('overseer.constants')
    local task_list = require('overseer.task_list')
    local util = require('overseer.util')
    local STATUS = constants.STATUS

    local symbols = {
      [STATUS.FAILURE] = '󰅚 ',
      [STATUS.CANCELED] = ' ',
      [STATUS.SUCCESS] = '󰄴 ',
      [STATUS.RUNNING] = '󰑮 ',
    }

    return function()
      local tasks = task_list.list_tasks()
      local tasks_by_status = util.tbl_group_by(tasks, 'status')
      local pieces = { '󰍹 ' }
      for _, status in ipairs(STATUS.values) do
        local status_tasks = tasks_by_status[status]
        if symbols[status] and status_tasks then table.insert(pieces, symbols[status] .. #status_tasks) end
      end
      return table.concat(pieces, ' ')
    end
    ```
- Instead of a single `require('mini.statusline').setup()`, define custom active active function (using [this base example as reference](https://github.com/echasnovski/mini.nvim/blob/14f81b96d850d561b6ed0c1e6715fdd724e4447f/doc/mini-statusline.txt#L86-L106)) like this:
    ```lua
    -- Make sure that 'stevearc/overseer.nvim' is loaded and set up prior to this
    local overseer_status = require('overseer-stl')

    --stylua: ignore
    local active_content = function()
      local mode, mode_hl = MiniStatusline.section_mode({ trunc_width = 120 })
      local git           = MiniStatusline.section_git({ trunc_width = 40 })
      local diff          = MiniStatusline.section_diff({ trunc_width = 75 })
      local diagnostics   = MiniStatusline.section_diagnostics({ trunc_width = 75 })
      local lsp           = MiniStatusline.section_lsp({ trunc_width = 75 })
      local filename      = MiniStatusline.section_filename({ trunc_width = 140 })
      local fileinfo      = MiniStatusline.section_fileinfo({ trunc_width = 120 })
      local location      = MiniStatusline.section_location({ trunc_width = 75 })
      local search        = MiniStatusline.section_searchcount({ trunc_width = 75 })

      local overseer = overseer_status()

      return MiniStatusline.combine_groups({
        { hl = mode_hl,                  strings = { mode } },
        { hl = 'MiniStatuslineDevinfo',  strings = { git, diff, diagnostics, lsp, overseer } },
        '%<', -- Mark general truncate point
        { hl = 'MiniStatuslineFilename', strings = { filename } },
        '%=', -- End left alignment
        { hl = 'MiniStatuslineFileinfo', strings = { fileinfo } },
        { hl = mode_hl,                  strings = { search, location } },
      })
    end

    require('mini.statusline').setup({ content = { active = active_content } })
    ```

Does this answer your question?

**Reply by dhruvinsh** [reply]
Perfect, this works as expected. 

side question,
for `mini.statusline` is it possible to inject component on the fly?

like when overseer loads (in case of lazy loading), I will inject it's component

**Reply by echasnovski** [reply]
> side question, for `mini.statusline` is it possible to inject component on the fly?
> 
> like when overseer loads (in case of lazy loading), I will inject it's component

Sure, but needs some Lua software engineering. There are several approaches here:
- You can set `MiniStatusline.config.content.active` any time and it will take effect. So set it directly to different function when 'overseer.nvim' is loaded. With something like `MiniStatusline.config.content.active = f`.
- Make `overseer_status` global (or more precisely - not local) with initial value of `function() return '' end` and change it to the "proper" function after 'overseer.nvim' is loaded.
- Probably there are other approaches.

**Reply by dhruvinsh** [reply]
Any plan on exposing [H.default_content_active ](https://github.com/echasnovski/mini.statusline/blob/85014aa9e4afe9a3ff1896ad768bf09584bff574/lua/mini/statusline.lua#L606) from `mini.statusline` ?

That way I can re-use that function to do some table insertion at proper index when `overseer.nvim` get loaded.

**Reply by echasnovski** [reply]
> Any plan on exposing [H.default_content_active ](https://github.com/echasnovski/mini.statusline/blob/85014aa9e4afe9a3ff1896ad768bf09584bff574/lua/mini/statusline.lua#L606) from `mini.statusline` ?
> 
> That way I can re-use that function to do some table insertion at proper index when `overseer.nvim` get loaded.

I am afraid no. It is [documented](https://github.com/echasnovski/mini.nvim/blob/14f81b96d850d561b6ed0c1e6715fdd724e4447f/doc/mini-statusline.txt#L86-L106) and this should be enough.

**Comment by dhruvinsh** [comment]
Thank you @echasnovski, my current solution with your help is,

- configure `mini.statuline` as I like
- when I load `overseer.nvim` modify the active content of `statusline`

```lua
-- overseer.lua config with lazy.nvim
return {
  "stevearc/overseer.nvim",
  cmd = { "Build", "OverseerClose", "OverseerOpen", "OverseerRun", "OverseerToggle" },
  opts = {
    templates = { "builtin" },
    task_list = {
      default_detail = 2,
      direction = "bottom",
      max_width = { 600, 0.7 },
    },
  },
  config = function(_, opts)
    local overseer = require("overseer")
    overseer.setup(opts)

    -- mini.statusline config

    local constants = require("overseer.constants")
    local task_list = require("overseer.task_list")
    local util = require("overseer.util")
    local STATUS = constants.STATUS

    local symbols = {
      [STATUS.FAILURE] = "󰅚 ",
      [STATUS.CANCELED] = " ",
      [STATUS.SUCCESS] = "󰄴 ",
      [STATUS.RUNNING] = "󰑮 ",
    }

    local overseer_status = function()
      local tasks = task_list.list_tasks()
      local tasks_by_status = util.tbl_group_by(tasks, "status")
      local pieces = { "󰍹 " }
      for _, status in ipairs(STATUS.values) do
        local status_tasks = tasks_by_status[status]
        if symbols[status] and status_tasks then
          table.insert(pieces, symbols[status] .. #status_tasks)
        end
      end
      return table.concat(pieces, " ")
    end
      
      --stylua: ignore
      local active_content = function()
        local mode, mode_hl = MiniStatusline.section_mode({ trunc_width = 120 })
        local git           = MiniStatusline.section_git({ trunc_width = 40 })
        local diff          = MiniStatusline.section_diff({ trunc_width = 75 })
        local diagnostics   = MiniStatusline.section_diagnostics({ trunc_width = 75 })
        local lsp           = MiniStatusline.section_lsp({ trunc_width = 75 })
        local filename      = MiniStatusline.section_filename({ trunc_width = 140 })
        local fileinfo      = MiniStatusline.section_fileinfo({ trunc_width = 120 })
        local location      = MiniStatusline.section_location({ trunc_width = 75 })
        local search        = MiniStatusline.section_searchcount({ trunc_width = 75 })

        local overseer = overseer_status()

        return MiniStatusline.combine_groups({
          { hl = mode_hl,                  strings = { mode } },
          { hl = 'MiniStatuslineDevinfo',  strings = { git, diff, diagnostics, lsp, overseer } },
          '%<', -- Mark general truncate point
          { hl = 'MiniStatuslineFilename', strings = { filename } },
          '%=', -- End left alignment
          { hl = 'MiniStatuslineFileinfo', strings = { fileinfo } },
          { hl = mode_hl,                  strings = { search, location } },
        })
      end

    MiniStatusline.config.content.active = active_content
  end,
}


```

**Reply by echasnovski** [reply]
> Thank you @echasnovski, my current solution with your help is,
> 
>     * configure `mini.statuline` as I like
> 
>     * when I load `overseer.nvim` modify the active content of `statusline`

Great that it works. Be aware also that this approach is not easily scalable if you want another custom component added.

**Reply by dhruvinsh** [reply]
> Be aware also that this approach is not easily scalable if you want another custom component added.

Agree, at that point I will extract the `active_content` which returns the default config (and that is why asking for exposing) and inject all the component that I need.

I just need to take care of `combine_groups` as it returns string not table. 

- get default component as `table`
- inject custom component
- use `combine_groups` to convert back to string.

**Reply by echasnovski** [reply]
This all can be achieved inside user's config. The structure of the default statusline is documented as a function. Anything beyond that is left for users to decide how they want to use it.

This was chosen as the most flexible approach to customizing statusline content (albeit at the cost of a big increase in line count for a first small change).

**Comment by dhruvinsh** [comment]
After using this for a while found out that there is one issue. There needs to be some type of continuous evaluation for task else if status set to running always stay as `running`.

Something that lualine component [has](https://github.com/stevearc/overseer.nvim/blob/236e60cdac6410dd95ea5cecafdb801a304d6a41/lua/lualine/components/overseer.lua#L91-L112) on overseer source code.

**Reply by echasnovski** [reply]
Not sure what the issue is here.

The output of 'lua/overseer-stl.lua' from [this comment](https://github.com/echasnovski/mini.nvim/discussions/1200#discussioncomment-10559736) is almost directly a `M.update_status()` from the linked code. And I've tested the component before posting: running tasks did end up showing as finished.

**Reply by dhruvinsh** [reply]
> running tasks did end up showing as finished.

Thank you for confirming. I will try to debug, might be my config have some issue, from [this comment](https://github.com/echasnovski/mini.nvim/discussions/1200#discussioncomment-10561667)
